### PR TITLE
Pytorch and Lightning 2.0 fixes

### PR DESCRIPTION
**Pull request by shrinath-suresh**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Pytorch 2.0 and Lightning 2.0 has been released recently.

Lightning 2.0 changes are published [here](https://lightning.ai/pages/releases/2.0.0/)

Following are the impacts on mlflow pytorch library

1. The package name has been changed from `pytorch_lightning` to `lightning`
2. `validation_epoch_end` and `test_epoch_end` has been removed
3. `pl.Trainer.add_argparse_args(parent_parser=parser)` has been removed. This line of code combines all the pytorch lightning command line arguments with the user's ArgumentParser. Instead we have to use LightningCLI as mentioned in the release link
4. The training arguments have changed. Previously, we use `--max_epochs` . Now, its `--trainer.max_epochs`
5. When installing `lightning` package, `pytorch_lightning` gets installed as well. Hence, we dont have to change everything in `mllfow.pytorch.autolog` library. But, we may need to remove older version checks.

## How is this patch tested?

- MNIST
[mnist_cpu_conda.txt](https://github.com/mlflow/mlflow/files/11032318/mnist_cpu_conda.txt)
[mnist_cpu_virtualenv.txt](https://github.com/mlflow/mlflow/files/11032319/mnist_cpu_virtualenv.txt)
[mnist_gpu_conda.txt](https://github.com/mlflow/mlflow/files/11043544/mnist_gpu_conda.txt)

- BERT
[bert_cpu_conda .txt](https://github.com/mlflow/mlflow/files/11032316/bert_cpu_conda.txt)
[bert_gpu_conda.txt](https://github.com/mlflow/mlflow/files/11043843/bert_gpu_conda.txt)

- Captum
[captum_cpu_virtual_env.txt](https://github.com/mlflow/mlflow/files/11032317/captum_cpu_virtual_env.txt)

- AX
[ax_hyper_param_tuning_cpu_conda.txt](https://github.com/mlflow/mlflow/files/11039550/ax_hyper_param_tuning_cpu_conda.txt)

- Iterative Pruning
[iterative_pruning_cpu_conda.txt](https://github.com/mlflow/mlflow/files/11043805/iterative_pruning_cpu_conda.txt)

Observation:
In multi gpu tests, the lightning throws `_thread.lock` object error while pickling the model and the model is not getting logged into mlflow. However, the issue has been fixed in lightning master branch. Attaching screenshot for reference

[screenshots.zip](https://github.com/mlflow/mlflow/files/11043814/screenshots.zip)

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
@shrinath-suresh thanks for the PR, ping me when it's ready for review!

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh thanks for the PR, ping me when it's ready for review!

sure @harupy . We are working on the GPU tests. Will let you know once the PR is ready for review

File: mlflow/pytorch/__init__.py [hunk 1]
```diff
@@ -1084,13 +1084,16 @@ def print_auto_logged_info(r):
...
     try:
-        import pytorch_lightning as pl
+        import lightning as L
     except ImportError:
-        pass
-    else:
+        import pytorch_lightning as pl
         from mlflow.pytorch._lightning_autolog import patched_fit
 
         safe_patch(FLAVOR_NAME, pl.Trainer, "fit", patched_fit, manage_run=True)
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:1092]: Can we do the following?

```python
    try:
        import pytorch_lightning as pl
    ...

    try:
        import lightning as L
    ...
```

Comment (shrinath-suresh) [reply mlflow/pytorch/__init__.py:1092]: the `lightning` package internally installs `pytorch_lightning`. The import `import pytorch_lightning as pl` still works in 2.0. if we set the code this way, `safe_patch` would get executed twice.

**Comment by shrinath-suresh** [comment]
@harupy Observing the following error message in CI - `Message: 'Could not install packages due to an OSError: [Errno 28] No space left on device\n'` ... Some of the test cases are failing during the package installation.

And the cross version tests are failing.. When i check the failure logs.. it has failed in `import pytorch_lightning as pl` . And in the master branch autolog code, the import is in try except block with `pass` in except - [code link](https://github.com/mlflow/mlflow/blob/1aa5434e6643de9b5d2ce78b095d42900bcde1ca/mlflow/pytorch/__init__.py#L1089)

For running autolog test, `pytorch_lightning` or `lightning` is mandatory right ?

**Comment by harupy** [comment]
> Message: 'Could not install packages due to an OSError: [Errno 28] No space left on device\n'

Where did you see this error? Can you share the URL?

> For running autolog test, pytorch_lightning or lightning is mandatory right ?

No, it's not mandatory. See https://github.com/mlflow/mlflow/pull/7627 for more details.

**Comment by shrinath-suresh** [comment]
> > Message: 'Could not install packages due to an OSError: [Errno 28] No space left on device\n'
> 
> Where did you see this error? Can you share the URL?
> 
> > For running autolog test, pytorch_lightning or lightning is mandatory right ?
> 
> No, it's not mandatory. See #7627 for more details.

I was getting the Space error in this [run](https://github.com/mlflow/mlflow/actions/runs/4488931479/jobs/7894121505) -  . But in the latest run i am not seeing it.

Thanks for the reference. Cross functional tests (for pytorch flavor) failed  due to the import error.. i fixed it supporting all three options (lightning, pytorch lightning <=1.9.4 and pytorch flavor)

```
    # For lightning>=2.0.0 flavor
    try:
        import lightning as L
    except ImportError:
        # for pytorch-lightning <= 1.9.4 flavor
        try:
            import pytorch_lightning as pl
        except ImportError:
            # For PyTorch flavor
            pass
        else:
            from mlflow.pytorch._lightning_autolog import patched_fit

            safe_patch(FLAVOR_NAME, pl.Trainer, "fit", patched_fit, manage_run=True)
    else:
        from mlflow.pytorch._lightning_autolog import patched_fit

        safe_patch(FLAVOR_NAME, L.Trainer, "fit", patched_fit, manage_run=True)
```

**Comment by harupy** [comment]
@shrinath-suresh What happens if we patch both `L.Trainer` and `pl.Trainer`?

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh What happens if we patch both `L.Trainer` and `pl.Trainer`?

I haven't tested the case where we patch both of it. Can you please explain a bit more on the context. 

As of now, the logic is to patch `L.Trainer` for lightning package and `pl.Trainer` for the ptl<=1.9.4. If we patch both, ptl<=1.9.4 will not have `L.Trainer` and fail.

**Comment by harupy** [comment]
> I haven't tested the case where we patch both of it. Can you please explain a bit more on the context.

Sure. If patching both Trainer is fine, we can simplify the code as I suggested.

```python
    try:
        import pytorch_lightning as pl
    except ImportError:
        pass
    else:
        from mlflow.pytorch._lightning_autolog import patched_fit

        safe_patch(FLAVOR_NAME, pl.Trainer, "fit", patched_fit, manage_run=True)

    try:
        import lightning as L
    except ImportError:
        pass
    else:
        from mlflow.pytorch._lightning_autolog import patched_fit

        safe_patch(FLAVOR_NAME, L.Trainer, "fit", patched_fit, manage_run=True)
```

**Comment by harupy** [comment]
Thanks for the update, let's see what happens.

**Comment by shrinath-suresh** [comment]
> Thanks for the update, let's see what happens.

Thanks for the suggestion @harupy . I added both and tested it in cpu and gpu.. Can you please review and let me know your comments.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**